### PR TITLE
IGNITE-17369 : v2, notify snapshot manager.

### PR DESCRIPTION
**Pull request by Vladsz83**
Solution: nitify, "ping" snapshot manager on every dengerous update.

### CODE REVIEW

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerTest.java [hunk 1]
```diff
@@ -3061,6 +3071,71 @@ public void testClusterSnapshotCreateSynchronously() throws Exception {
...
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testClusterCreateSnapshotWarning() throws Exception {
+        IgniteConfiguration cfg = getConfiguration(getTestIgniteInstanceName(0));
+        cfg.getConnectorConfiguration().setHost("localhost");
+
+        String targetMsg = U.field(DataStreamerUpdatesHandler.class, "WRN_MSG");
+
+        AtomicReference<Exception> simulationEx = new AtomicReference<>();
```
Comment (NSAmelchev) [inline modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerTest.java:3086]: Why do we need it?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteClusterShanpshotStreamerTest.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+    /**
+     * Tests snapshot process throws warning if required.
+     *
+     * @param mustFail If {@code true}, checks snapshot process produces the warning. Otherwise, checks snapshot
+     *                 process produces no warning.
+     * @param allowOverwrite 'allowOverwrite' setting.
+     */
+    private void doTestDataStreamerWhileSnapshot(boolean mustFail, boolean allowOverwrite) throws Exception {
+        String expectedWrn = U.field(DataStreamerUpdatesHandler.class, "WRN_MSG");
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteClusterShanpshotStreamerTest.java:187]: Lets change a modifier of the `WRN_MSG` to a public.

File: modules/core/src/test/java/org/apache/ignite/testsuites/IgniteSnapshotTestSuite.java [hunk 3]
```diff
@@ -46,7 +47,8 @@
     IgniteSnapshotRestoreFromRemoteTest.class,
     PlainSnapshotTest.class,
     EncryptedSnapshotTest.class,
-    IgniteClusterSnapshotWalRecordTest.class
+    IgniteClusterSnapshotWalRecordTest.class,
+    IgniteClusterShanpshotStreamerTest.class
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/testsuites/IgniteSnapshotTestSuite.java:51]: typo: Shanpshot

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.ignite.internal.util.typedef.F;
+
+import static org.apache.ignite.internal.processors.cache.persistence.snapshot.SnapshotHandlerType.CREATE;
+
+/**
+ * A snapshot haldler that monitors and warns of concurrent DataStreamer updates.
+ */
+public class DataStreamerUpdatesHandler implements SnapshotHandler<Boolean> {
+    /** */
+    private static final String WRN_MSG = "DataStreamer with property 'alowOverwrite' set to `false` was working " +
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java:33]: typo: alowOverwrite

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    @Override public Boolean invoke(SnapshotHandlerContext ctx) {
+        return ctx.streamerWarning();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void complete(String name, Collection<SnapshotHandlerResult<Boolean>> results)
+        throws SnapshotHandlerWarningException {
+        Collection<UUID> nodes = F.viewReadOnly(results, r -> r.node().id(), SnapshotHandlerResult::data);
+
+        if (!nodes.isEmpty())
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java:53]: Brackets {} should be here

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotHandlerContext.java [hunk 6]
```diff
@@ -38,17 +38,24 @@ public class SnapshotHandlerContext {
...
      * @param metadata Snapshot metadata.
      * @param grps The names of the cache groups on which the operation is performed.
+     * @param streamerWrn {@code True} if concurrent streaming updates occured during snapshot operation.
+     * {@code False} otherwise. Always {@code false} for snapshot restoration.
      * @param locNode Local node.
      * @param snpDir The full path to the snapshot files.
      */
-    public SnapshotHandlerContext(SnapshotMetadata metadata, @Nullable Collection<String> grps, ClusterNode locNode, File snpDir) {
+    public SnapshotHandlerContext(SnapshotMetadata metadata, @Nullable Collection<String> grps,
+        boolean streamerWrn, ClusterNode locNode, File snpDir) {
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotHandlerContext.java:53]: Lets make `streamerWrn` parameter last.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.Collection;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import org.apache.ignite.internal.util.typedef.F;
+
+import static org.apache.ignite.internal.processors.cache.persistence.snapshot.SnapshotHandlerType.CREATE;
+
+/**
+ * A snapshot haldler that monitors and warns of concurrent DataStreamer updates.
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java:29]: Lets add that it means operations that may lead to inconsistent snapshot create operation.

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerTest.java (hunk 1 above)
Comment (Vladsz83) [reply modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerTest.java:3086]: Removed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteClusterShanpshotStreamerTest.java (hunk 2 above)
Comment (Vladsz83) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteClusterShanpshotStreamerTest.java:187]: Fixed

File: modules/core/src/test/java/org/apache/ignite/testsuites/IgniteSnapshotTestSuite.java (hunk 3 above)
Comment (Vladsz83) [reply modules/core/src/test/java/org/apache/ignite/testsuites/IgniteSnapshotTestSuite.java:51]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java (hunk 4 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java:33]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java (hunk 5 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java:53]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotHandlerContext.java (hunk 6 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotHandlerContext.java:53]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java (hunk 7 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/DataStreamerUpdatesHandler.java:29]: Fixed

**Review by NSAmelchev — APPROVED** [review]

**Pull request merged**